airgun/views/redhat_repository.py

```diff
@@ -0,0 +1,269 @@
+    ActionsDropdown,
+)
+
+
+def _wait_for_spinner(widget):
+    """Wait for any spinner to disappear from widget"""
+    wait_for(
+        lambda: widget.is_displayed and not widget.browser.elements(
+            ".//div[contains(@class, 'spinner')]", parent=widget),
+        timeout=60,
+        delay=1,
+        logger=widget.logger
+    )
+
+
+class AvailableRepositoryItem(GenericLocatorWidget):
+    """The widget representation of Available repository item of an Available repository set."""
+    ENABLE_BUTTON = './/button'
+    TEXT = './/span'
+
+    @property
+    def text(self):
+        """Return the text representation of this repository eg: architecture + OS version if
+        available.
+        """
+        return self.browser.text(self.TEXT, parent=self)
+
+    def read(self):
+        """Read this widget by return it's text representation."""
+        return self.text
+
+    def enable(self):
+        """Enable this repository."""
+        self.browser.click(self.ENABLE_BUTTON, parent=self)
+        _wait_for_spinner(self.parent)
+
+    def fill(self, value):
+        """Fill this widget by enabling the repository if value is True"""
+        if value is True:
+            self.enable()
+
+
+class AvailableRepositorySetWidget(GenericLocatorWidget):
+    """The widget representation of Available repository set item."""
+    ITEM = AvailableRepositoryItem
+    EXPAND_BUTTON = ".//div[contains(@class, 'expand')]"
+    NAME = ".//div[contains(@class, 'item-heading')]"
+    LABEL = ".//div[contains(@class, 'item-text')]"
+    ITEMS = ".//div[contains(@class, 'list-item-with-divider')]"
+
+    @property
+    def expand_button(self):
+        """Return the expand button element."""
+        return self.browser.element(self.EXPAND_BUTTON, parent=self)
+
+    @property
+    def expanded(self):
+        """Check whether this repository set is expanded or not."""
+        return 'active' in self.browser.get_attribute('class', self.expand_button)
+
+    def expand(self):
+        """Expand the repository set item section."""
+        if not self.expanded:
+            self.browser.click(self.EXPAND_BUTTON, parent=self)
+            _wait_for_spinner(self.parent)
+
+    @property
+    def name(self):
+        """Return the name displayed for this repository item"""
+        return self.browser.text(self.NAME, parent=self)
+
+    @property
+    def label(self):
+        """Return the label displayed for this repository item"""
+        return self.browser.text(self.LABEL, parent=self)
+
+    @property
+    def items(self):
+        """Return all the items (available repositories) of this repository set."""
+        if self.ITEM:
+            self.expand()
+            return [
+                AvailableRepositoryItem(self, '{0}[{1}]'.format(self.ITEMS, index + 1))
+                for index, _ in enumerate(self.browser.elements(self.ITEMS, parent=self))
+            ]
+        return []
+
+    def read_items(self):
```

Comment:
why do we need both read and read_items? Why don't we combine them into one?

Reply:
The read will read only the name and label to read the repo set items we have to expand the repository set that will make a server call to get the items, it's same way that when we have a table we read only the table items of the the entity but to read the entity we have to open it and read the edit form , I'am using the same analogy as for simple entity, as if we have 20 entities in the table we are not going to open each entity and read it.

Reply:
EG: we need this only when we are reading a concrete defined entity.